### ISSUE TITLE
A better way to work w Epics

### ISSUE THREAD

**Issue by andycleff** [ZenHubIO/support#424]
As a scrum master, product owner, or dev-team member, I would like to be able to create Epics, add/remove individual issues to the Epics, so that I have better visibility of progress of issues contained therein. 

Notes: I have no idea of the constraints that GitHub imposes that might impact this idea.

**Comment by goanpeca** [comment]
I second this, what we are doing right now is create an issue with a label `Epic` and we use this issue to aggregate the other issues. 

With something like this it would be nice to have a summary of estimates from all the issues referenced in this "epic".

Also if github allowed for more customization of issue links, we could do something like

```
#456{title}
```
and get something like
[#456 The title of the issue]()

**Comment by andycleff** [comment]
we are doing the same, label epic and a check box list of the related issues 
[ ] and [x]

We do get some primitive indication via git of how many check boxes have been completed.

![screenshot 2015-11-05 08 14 51](https://cloud.githubusercontent.com/assets/7416277/10969292/5ba95140-8395-11e5-9428-acda08d706bb.png)

Perhaps adding that to the ZenHub card display?

Current:
![screenshot 2015-11-05 08 15 55](https://cloud.githubusercontent.com/assets/7416277/10969386/fc05704c-8395-11e5-91ca-48691a49dfd6.png)


with Additional info:
![epic display](https://cloud.githubusercontent.com/assets/7416277/10969379/f6091428-8395-11e5-97ba-fbc5801aa33f.jpg)

**Comment by goanpeca** [comment]
Indeed!, and also add all the points for the included issues!

**Comment by andycleff** [comment]
would you add those points into the Epics points? Or have a separate indication of issue story points?

**Comment by goanpeca** [comment]
I think Epics should not have assigned directly. An epic should calculate the points based on the issues it contains...

**Comment by benschac** [comment]
Would it be possible to integrate with a product like this or have an epic feature that's similar? 
http://sprinthub.io/

**Comment by andycleff** [comment]
The terminology that sprinthub.io is choosing confuses me..... seems very non-standard... anyone else feel the same way?

Regardless, their idea of finding a better way to collect and manage multiple git issues into some sort of hierarchy is good.... 

In parenthesis are what i think are github equivalents

Project (repo) > Releases (milestones) > Features (labels) > Epics (issue) > Stories (issue) > Tasks (issue)

It would be really nice to start w a better way to collect git issues into epics... the UI that sprinthub provides might be one way that is kinda sorta close....

See: http://sprinthub.io/ - what they call a "story" would be a nice way to handle what I consider an "epic"

Sprinthub's use of "assigned/unassigned" is confusing... I think they don't mean assigned in the context of an individual developer being assigned to an issue (standard Git use of the term.) Instead I think they mean a story is not tagged with a milestones.

ZenHub - you guys can do much better!!!!

**Comment by azenMatt** [comment]
Thank for all the feedback here. 

I completely agree that a way to do epics would be very useful - and I get plenty of feedback from people who agree :) 

We have been looking into different solutions to create these that still fit in with the GitHub experience. 

@andycleff - I really like your idea of including the progress bar on the card: 

![image](https://cloud.githubusercontent.com/assets/7584089/11161347/0accb366-8a2f-11e5-96a4-c012ab1b6c0a.png)

which has been discussed here - #197 - and could be a great way to help create a visualize difference between 'normal' issue cards and 'epic' cards. 

Perhaps a way to be able to select these on the board and have all non-related issues (i.e those not on the checklist) disappear would also be useful?  




> would you add those points into the Epics points? Or have a separate indication of issue story points?

Using the method above, we could leave this up to the user, by putting a sum total here: 

![image](https://cloud.githubusercontent.com/assets/7584089/11161380/db263578-8a2f-11e5-8731-0e360d21e298.png)

for example `10/25` - to show the number of story points completed vs total story points. You can decide if you want to use this figure or override it with your own estimation.

**Comment by goanpeca** [comment]
:+1:

**Comment by stephenakearns** [comment]
I'm supportive of the idea. One additional observation: The underlying challenge to me seems that there's no hierarchical construct in GitHub Issues. (e.g. I can organize/group labels).  I can't group issues/stories into an epic. I don't know if there's an easy way to solve this or not.

**Comment by goanpeca** [comment]
Maybe a bit more complex... but the same way we use Epics, we could use stories...

- A normal issue represents... an issue
- An issue with a **Story** label would represent a story and would group a list of issues, the sum of the story could be set manually or calculated by the sum of the included issues
- An issue with an **Epic** label would represent an epic and could have stories included inside.... the sum would be manual or also calculated by the sum of the child stories (and the respective child issues)

The next step would be to include a new visualization tool that would allow some grouping of the Epics, and a drop-down or accordion or something that would display the contained issues 

This is what Jira offers,... (on the left)

![image](https://cloud.githubusercontent.com/assets/3627835/11162120/c7f83de4-8a60-11e5-9e71-0dbdccfbc245.png)

**Comment by andycleff** [comment]
If  you want jira features, purchase jira <[;-)

**Comment by andycleff** [comment]
ZenHub... got any love coming for this ticket? Start w low hanging fruit perhaps: 
![image](https://cloud.githubusercontent.com/assets/7416277/11757179/4a997dd0-a02d-11e5-86f3-e22f317ca6f6.png)

**Comment by azenMatt** [comment]
Yes :smile: 

i'm a big fan of this, and love the ideas in this thread. Expect to see some movement (early) next year

**Comment by ljm22** [comment]
@azenMatt any movement on this?

Agree with @stephenakearns that there is a structural issue around lack of dependency within GitHub issues. But then, we don't want it to turn into Jira. Lord, no. 'Only as complicated as it needs to be' is our watchword. Watchphrase.

We're currently using the checklist approach to constructing an Epic, and have a routine that looks for issues that start with 'Epic: ' and uses the checklist to autogenerate issues with 'depends on #[epic issue number]' dropped into the description to provide open/closed notifications in the parent epic issue comments. Then it edits the Epic description to insert the approriate issue number agaiunst the checkbox. Yes, a bit hacky, but had to work within capability of the GitHub api. Having the progress bar on the card would be very nice indeed (and not just for epics). Presently we end up using a filtered view of the issues list. We keep the epic checkboxes up to date by hand (which is working ok so far).

What we haven't worked out is where in the pipeline an Epic should sit. I'm nopt even sure that the concept of 'pipeline' is useful for an epic in a Kanban environment. Epic as collapsible horizontal swimlane would be more useful.

We tried using milestones to group issues into epics but that resulted in unhelpful milestone proliferation.

We tried having an 'Epic' pipeline stage to hold Epic issues in but that's semantically weird.

An unforseen consequence of messing with the idea of Epics is the extra issues it's generating -- the Epic issues themselves and the subtasks that may have once just sat within a single issue before and are now turning into separate issues. Chrome is then struggling just to load the cards into the board view.

**Comment by azenMatt** [comment]
> But then, we don't want it to turn into Jira. Lord, no. 'Only as complicated as it needs to be' is our watchword. Watchphrase.

:100: I am very much in agreement with this! 


> any movement on this?

This is being actively worked on at the moment, targeting a release at the start of April. Based on your feedback, what we have should work very well for you! 



> An unforeseen consequence of messing with the idea of Epics is the extra issues it's generating -- the Epic issues themselves and the subtasks that may have once just sat within a single issue before and are now turning into separate issues. Chrome is then struggling just to load the cards into the board view.

This is also something that we are working on in parallel, so expect to see some performance improvements soon that will resolve this problem.

**Comment by andycleff** [comment]
\\0//

**Comment by matejkastner** [comment]
Hallo @azenMatt do I understand it right, that this task is currently live?

![image](https://cloud.githubusercontent.com/assets/10279303/14247338/cdace49c-fa6d-11e5-8083-f985b4ccde5a.png)

**Comment by andycleff** [comment]
.... and where would that lovely button be? I'm on ZenHub 2.16.11... not seeing it just yet

**Comment by andycleff** [comment]
ahh.... found it. Click "new issue" first then you have the option:

![screen shot 2016-04-04 at 8 20 06 am](https://cloud.githubusercontent.com/assets/7416277/14247712/17642864-fa3e-11e5-95d5-195c6fb6f94e.png)

**Comment by andycleff** [comment]
@azenMatt very nicely done!!!

**Comment by andycleff** [comment]
@azenMatt one very minor UI change suggestion. Change the button label from "Create Epic" to either "Save Epic" or "Update Epic"

As a user, I was very confused by the "Create Epic" label.... as I had just already created it... wasn't sure why I'd need to "create" again... would it duplicate things... etc.

![boards_ _dramafever_project-management](https://cloud.githubusercontent.com/assets/7416277/14248022/3d780398-fa40-11e5-9228-4ad64072c511.png)

**Comment by junketjim** [comment]
It took me a few moments but this feature is, 'scuse pun, "epic".  The ability to group issues from across repos. is essential for our teams (if it's not obvious to folk, you have to have repos. merged on a board before they are available to select for your epic).  Thank you ZenHub (PS. if you start going down the Jira-functionality rabbit hole I will cry)

**Comment by andycleff** [comment]
wait, you can do cross repo? I didn't see that yet... /me looking now

**Comment by andycleff** [comment]
awesome sauce! 
![screen_shot_2016-04-04_at_10_44_16_am](https://cloud.githubusercontent.com/assets/7416277/14251750/6da84728-fa52-11e5-8568-6a0168161da7.png)

**Comment by azenMatt** [comment]
Hi all! 

Congratulations on being the first people to find Epics in ZenHub! :tada: :smile: 

We launched this last night to give it time to get through to everyone browsers before we sent out an email update for it - which you should be getting about now. 



> one very minor UI change suggestion. Change the button label from "Create Epic" to either "Save Epic" or "Update Epic"

Good idea, we have a couple of other edges that need to be polished this week that we didn't want to delay the release for - so I will add that one to the list! 



> The ability to group issues from across repos. is essential for our teams (if it's not obvious to folk, you have to have repos. merged on a board before they are available to select for your epic)

Thanks @junketjim - as an aside, making this idea of merging repos together more obvious is something that is on our roadmap. At the moment it is very 'board based' which made sense for early versions of ZenHub. But now with more reporting and Epics also using the same shared repos, it is less obvious that the board should be the home for this. So we will be looking into how we can make the idea of merged repos more 'ZenHub wide'



> PS. if you start going down the Jira-functionality rabbit hole I will cry

I will join you! 




We are going to be making a few tweaks this week based on feedback, so please let me know your thoughts once you have had a chance to use it so we can wrap them in :smile:

**Comment by aspushkinus** [comment]
@azenMatt I'm not seeing it anywhere.

Things tried:
- Checking browser for updates (Chrome Version 49.0.2623.110 (64-bit));
- Restarting browser;
- Refreshing github feverishly;
- Wiping display with soft cloth.

**Comment by aspushkinus** [comment]
Sorry, never mind: re-installing Zenhub extension fixed the issue! :)

**Comment by azenMatt** [comment]
Glad you found it @aspushkinus :) we are looking into better ways to 'go live' with big releases like this as we are currently dependant on your browser checking for updates. The default settings are Chrome checks every 5 hours and Firefox every 12, so sometimes it takes a while!

**Comment by andycleff** [comment]
Can always force Chrome to update extensions, instead of uninstalling/resinstalling

First check developer mode, Preferences > Extensions
![screen shot 2016-04-04 at 2 13 48 pm](https://cloud.githubusercontent.com/assets/7416277/14258239/85b0cdaa-fa6f-11e5-8a6a-d17489caa11a.png)

Then update extensions:
![screen shot 2016-04-04 at 2 13 56 pm](https://cloud.githubusercontent.com/assets/7416277/14258249/99caa9e6-fa6f-11e5-958e-51f4b8ed3566.png)

**Comment by matejkastner** [comment]
Hallo @azenMatt,
best update after introducing time estimates! YES! Thank you so much!

We have been just before switching our epics to be planned in JIRA or other 3rd party tool, but did not switched yet since I did not wanted to feel the pain of all our developers being living in other world.

**Comment by ljm22** [comment]
The epics are, indeed, epic. Well done @azenMatt and team. Looks like some
speed improvements too?

On Mon, Apr 4, 2016 at 10:00 PM, Matej Kastner <notifications@github.com>
wrote:

> Hallo @azenMatt <https://github.com/azenMatt>,
> best update after introducing time estimates! YES! Thank you so much!
>
> We have been just before switching our epics to be planned in JIRA or
> other 3rd party tool, but did not switched yet since I did not wanted to
> feel the pain of all our developers being living in other world.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/ZenHubIO/support/issues/424#issuecomment-205494860>
>



-- 
Liam McGee
Taxonomics.co.uk
+(0)1373 813414

**Comment by ljm22** [comment]
And we really like the fact that an epic can be a sub-epic of another epic.

**Comment by ljm22** [comment]
Hi @azenMatt - can we re-order issues within an Epic? I would like to be able to create Epic issues in any order, but present them within the Epic in an ordered sequence.

**Comment by andycleff** [comment]
@ljm22 interesting thought!

Can we collectively flesh out some more details, some use cases...

For example:
- Order by issue number (can't think of a use case for this though)
- Order manually (reposition in epic view)
- Order by rank (vertical position in status column)
- Order by status (done at the bottom)

Let's brainstorm before we "filter"...

**Comment by goanpeca** [comment]
Great addition. great work :-)

**Comment by azenMatt** [comment]
> can we re-order issues within an Epic? I would like to be able to create Epic issues in any order, but present them within the Epic in an ordered sequence. I can uncheck and re-check to change the order in the Epic Creation view, but that doesn't maintain it's ordering or carry across to the Epic 'issue' view, just keeps ordering in order of issue number.

Open to ideas here - we did play around with a couple of different options during design, but couldn't get anything that really clicked. I figured it would be best to release it and then see how everyone would like to approach it once it has been live for a while. 

One option we are looking to include this week is a 'view in board' button, to make it easy to jump to a board which is already filtered by this epic. 

This means the issue view becomes 'what needs to be done' and the board stays as 'this is the order in which it needs to be done'

**Comment by ljm22** [comment]
+1 to view in board... We've got some hundreds of issues across 11 repos and Beong able to jump from the epic to the board view of that epic would be wonderful. And easy, by the looks of it.

**Comment by Firesphere** [comment]
Creating a new epic is a bit clumsy, it could use some more descriptive way of how they're created. At first I didn't understand where my epic went. Turned out I didn't go all the way. The final create epic button is not in clear sight.

Also, sadly, you can't link a new issue to an existing epic when creating them, that's a minor annoyance.

**Comment by andycleff** [comment]
In my experience better to ship a usable v1 then wait for perfection.

Yep, there's a bit of clumsy. But miles beyond native github!

Did everyone see the How To post: https://www.zenhub.io/blog/working-with-epics-in-github/

**Comment by jugdemon** [comment]
I couldn't find an issue describing my problem, so I hope this is a good place to bring it up:

We have a lot of epics in our board and their sub-issues are even more numerous. We would love to have a feature where we can "collapse" epics such that only the epic itself is visible (and all its issues are hidden) and we can focus on non-epic items.

In filter-terms, we would like a filter which only shows epics or issues that are not part of any epic. It would be nice to be able to expand single epics, but even if we only can either collapse or expand all issues that would already be of great help.

**Comment by andycleff** [comment]
1. Filter to show only epics is easy, as "Epic" is simply a label.
2. Expanding a single epic (to show its issues is also easy using the "Filter by Epic" button on an epic.
3. Issues that are not part of an epic can be identified by going into a single epic and using "Add or modify issues in this epic" and looking for non-checked issues. A bit cumbersome with loads of epics true...

**Comment by jugdemon** [comment]
@andycleff Thank you for the suggestions, but it does not quite fit our bill. The thing we want to be able to view - so to speak - on the board is all "top level issues" and none of the sub-issues (i.e. epic-members). Your suggestions gather those issues, but they go against the kanban board philosophy, where we should get an easy overview. I think providing such a (combined) filter on the board-level shouldn't be too difficult to implement and would improve productivity for people managing the board.

**Comment by andycleff** [comment]
I'm not following your "spec" for combined view. What would that filter show? Epic "cards" only? or something else?

**Comment by jugdemon** [comment]
Sorry for being unclear. I try to describe it differently. Essentially, an epic and its subissues can be seen as tree of depth 1 with the epic being the root the subissues being the leaves. If we now see all the issues on the board we get to see roots and leaves at once. Filtering for epics lets us see the roots of such trees. However, issues that are not inside an epic inconveniently disappear as well. The stupid solution would be to make every stand-alone issue an epic that way filtering for epics would return what we want to see. However, that seems to be non-practical. What we would want to filter for is for issues that are marked as epic and all issues that are not contained in any epic at once.

That way we can see all groups of tasks that need to be accomplished. The reason we bring that up is that if you have 10+ epics and each has 3 to 10 issues inside you end up with a screen full of issues that can be summarized with the epics themselves. All the other issues that are not part of epics simply drown out in the subissues of all the epics. If we could filter away the leaves of all epic, we could see the the tasks at hand (without seeing subtasks described by epics).

I am not sure it is clearer now, but we realized that it hinders our workflow and we wondered whether we are alone with that and if not whether it could be added to zenhub in the medium term.

**Comment by azenMatt** [comment]
@jugdemon @andycleff - do you think this could work well with the mock ups in - https://github.com/ZenHubIO/support/issues/68#issuecomment-236068055

in that these could be a filter option in the search box rather than it's own dedicated filter? 

It is a common request for us to mark all epic sub issue with a label as well (which we will look to do soon), so you could fill this use case with a filter `issue = epic and issue not= sub task' 

this should give you all epic issues and all stand alone issues, but not issues assigned to an epic

cc @brianleung11-3

**Comment by jugdemon** [comment]
That kind of filter would definitely be the right kind of filter. I also think that marking subtasks is more sustainable than having to mark the non-epic tasks. I look forward to that upgrade.

**Comment by andycleff** [comment]
"not" would be a very powerful addition to ZH filtering !

**Comment by kendramoroz** [comment]
👋  Doing some cleaning of our community Board, and we've decided to close this issue as it's quite outdated. Please feel free if you're leveraging Epics today and feel there's an area of it that can be improved on, spin up a new issue with feedback and we'll be sure to follow-up! In the meantime, keep an eye out on `Priorities` and `In Progress` to stay tuned with what we're up to! 😄
